src/facade/dragonfly_connection.cc

```diff
@@ -724,11 +724,26 @@ auto Connection::ParseMemcache() -> ParserStatus {
       break;
     }
 
+    auto parse_value = [](std::string_view str, auto consumed,
+                          auto bytes_len) -> std::optional<std::string_view> {
+      std::string_view value = str.substr(consumed, bytes_len);
+      auto protocol_value = str.substr(consumed);
+      auto pos = protocol_value.find("\r\n");
+      DCHECK(pos != std::string::npos);
+      protocol_value = protocol_value.substr(0, pos);
+      return protocol_value.size() == value.size() ? value : std::optional<std::string_view>{};
+    };
+
     size_t total_len = consumed;
     if (MemcacheParser::IsStoreCmd(cmd.type)) {
       total_len += cmd.bytes_len + 2;
       if (io_buf_.InputLen() >= total_len) {
-        value = str.substr(consumed, cmd.bytes_len);
+        auto maybe_value = parse_value(str, consumed, cmd.bytes_len);
+        if (!maybe_value) {
+          builder->SendClientError("bad data chunk");
+          return OK;
+        }
```

Comment:
I suspect that parse_value is both wrong and inefficient: `\r\n` could be part of the blob data, in that case we will parse something we should not. In addition, calling find("\r\n") on potentially a 20MB blob is inefficient and unnecessary. 
What about this function?

```suggestion
        string_view maybe_value = str.substr(consumed);
        if (maybe_value[cmd.bytes_len] != '\r' || maybe_value[cmd.bytes_len != '\n') {
          builder->SendClientError("bad data chunk");
          return ERROR;
        }
       value = maybe_value.substr(0, cmd.bytes_len);
```